Sources/GRPC/GRPCWebToHTTP2ServerCodec.swift

```diff
@@ -0,0 +1,482 @@
+        return .completePromise(promise, .success(()))
+      }
+
+    case .closed:
+      return .completePromise(promise, .failure(GRPCError.AlreadyComplete()))
+
+    case ._modifying:
+      preconditionFailure("Left in modifying state")
+    }
+  }
+}
+
+// MARK: - Helpers
+
+extension GRPCWebToHTTP2ServerCodec {
+  private static func makeResponseHead(hpackHeaders: HPACKHeaders) -> HTTPResponseHead {
+    let headers = HTTPHeaders(hpackHeaders: hpackHeaders)
+
+    // Grab the status, if this is missing we've messed up in another handler.
+    guard let statusCode = hpackHeaders.first(name: ":status").flatMap(Int.init) else {
+      preconditionFailure("Invalid state: missing ':status' pseudo header")
+    }
+
+    return HTTPResponseHead(
+      version: .init(major: 1, minor: 1),
+      status: .init(statusCode: statusCode),
+      headers: headers
+    )
+  }
+
+  private static func formatTrailers(
+    _ trailers: HPACKHeaders,
+    allocator: ByteBufferAllocator
+  ) -> ByteBuffer {
+    // See: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md
+    let encodedTrailers = trailers.map { name, value, _ in
+      "\(name): \(value)"
+    }.joined(separator: "\r\n")
+
+    var buffer = allocator.buffer(capacity: 5 + encodedTrailers.utf8.count)
+    // Uncompressed trailer byte.
+    buffer.writeInteger(UInt8(0x80))
+    // Length.
+    buffer.writeInteger(UInt32(encodedTrailers.utf8.count))
+    // Uncompressed trailers.
+    buffer.writeString(encodedTrailers)
+
+    return buffer
+  }
+
+  private static func encodeResponsesAndTrailers(
+    _ responses: inout CircularBuffer<ByteBuffer>,
+    trailers: HPACKHeaders,
+    allocator: ByteBufferAllocator
+  ) -> ByteBuffer {
+    // We need to encode the trailers along with any responses we're holding.
+    responses.append(self.formatTrailers(trailers, allocator: allocator))
+
+    let capacity = responses.lazy.map { $0.readableBytes }.reduce(0, +)
+    // '!' is fine: responses isn't empty, we just appended the trailers.
+    var buffer = responses.popFirst()!
+
+    // Accumulate all the buffers into a single 'Data'. Ideally we wouldn't copy back and forth
+    // but this is fine for now.
+    var accumulatedData = buffer.readData(length: buffer.readableBytes)!
+    accumulatedData.reserveCapacity(capacity)
+    while let buffer = responses.popFirst() {
+      accumulatedData.append(contentsOf: buffer.readableBytesView)
+    }
+
+    // We can reuse the popped buffer.
+    let base64Encoded = accumulatedData.base64EncodedString()
+    buffer.clear(minimumCapacity: base64Encoded.utf8.count)
+    buffer.writeString(base64Encoded)
+
+    return buffer
+  }
+}
+
+extension HTTPHeaders {
+  fileprivate init(hpackHeaders headers: HPACKHeaders) {
+    self.init()
+    self.reserveCapacity(headers.count)
+    // Don't add any pseudo-headers.
+    for (name, value, _) in headers where name.utf8.first != .some(UInt8(ascii: ":")) {
```

Comment:
There's a mild perf improvement available here: well-formatted HTTP/2 header blocks must have all their pseudo-headers _first_. You can therefore do a `dropFirst(where:)` to only look at the minimal number of headers and then just do `add(contentsOf:)` or similar to reduce the amount of unnecessary checking you do.

Reply:
Oh, smart thinking, I like it!